lang/src/bpf_upgradable.rs

```diff
@@ -0,0 +1,85 @@
+use crate::{AccountDeserialize, AccountSerialize, Owner};
+use solana_program::{
+    bpf_loader_upgradeable::{self, UpgradeableLoaderState},
+    program_error::ProgramError,
+    pubkey::Pubkey,
+};
+
+#[derive(Clone)]
+pub struct ProgramData {
```

Comment:
I don't have a great sense, but having both `ProgramData` and `UpgradeableLoaderState` at the account context level might be redundant. What do you think about always requiring the use of `Account<UpgradeableLoaderState>` and then calling `try_into()` to get the `ProgramData`? It's less convenient to call `try_into`, but it might be nice to minimize the number of types used in the context.

Reply:
I think in 99% of cases users want either the program_data or the program, not the buffer or unitialized variant. If we handle program state in the already existing Program struct like I described in the comment above, in most cases users want the ProgramData when they use Account< UpgradeableLoaderState> so it would sound good to me to make that flow easy.

I'm also not sure try_into() would work without some changes to `has_one`. If you want to check that another account equals the uprade_authority, currently it expects that both targets are the same but that wont work with `try_into()` so youd have to use a raw constraint. I have suggested a change that would make this possible with has_one though (https://github.com/project-serum/anchor/issues/1093). Although there's also the problem that Programdata's key is optional but I have an issue for that too (https://github.com/project-serum/anchor/issues/1092)

also not sure just `try_into()` would be enough, my guess is that it would need type annotation because it all happens in an if statement so it might have to be `try_into::<ProgramData>()`

so let's assume that the redirect and option feature I suggested get implemented then you have this

`#[account(has_opt = try_into::<ProgramData>().upgrade_authority_address -> authority)]`

vs

`[#account(has_opt = authority)]`

or without them

`#[account(constraint = me. try_into::<ProgramData>().upgrade_authority_address = Some(authority.key())`

vs

`[account(constraint = me.upgrade_authority_address = Some(authority.key())]`

without them it doesn't make that much of a difference but with the features, I think the former takes quite some more space and is more difficult to parse. doesnt feel as good to me for what I think will be the default use case. but yea, also depends on how we want to handle program state

Reply:
also, if I understand correctly, we would still have a `ProgramData` struct to `try_into()` (because it's just a variant in the state) so we'd have that type anyway

Reply:
Definitely agree on the above. ProgramData sgtm.

Reply:
could

what do we do with program state tho?

Do we a) use Account<'info, Program> or do we b) integrate state into the existing Program struct with an optional programdata_address field if it's a bpf_upgradable program or do we c) not do anything and let this case be handled by Account<info, UpgradeableLoaderState>?

i like b) the most cause it also does the check that it's the correct program and we already have that type